pkg/tbtc/signing.go

```diff
@@ -2,378 +2,372 @@ package tbtc
 	if err != nil {
-		return nil, fmt.Errorf("cannot get qualified operators: [%w]", err)
+		return nil, 0, fmt.Errorf("cannot marshal wallet public key: [%v]", err)
 	}
 
-	// Exclude all members controlled by the operators that were not
-	// qualified for the current attempt.
-	return srl.excludedMembersIndexes(
-		qualifiedOperatorsSet,
-		readyMembersIndexes,
-	), nil
-}
+	loopTimeoutBlock := startBlock +
+		uint64(se.signingAttemptsLimit*signingAttemptMaximumBlocks())
 
-// qualifiedOperatorsSet returns a set of operators qualified to participate
-// in the given signing attempt. The set of qualified operators is taken
-// from the set of active operators who announced readiness through
-// their controlled signing group members.
-func (srl *signingRetryLoop) qualifiedOperatorsSet(
-	readyMembersIndexes []group.MemberIndex,
-) (map[chain.Address]bool, error) {
-	// The retry algorithm expects that we count retries from 0. Since
-	// the first invocation of the algorithm will be for `attemptCounter == 1`
-	// we need to subtract one while determining the number of the given retry.
-	retryCount := srl.attemptCounter - 1
-
-	var readySigningGroupOperators []chain.Address
-	for _, memberIndex := range readyMembersIndexes {
-		readySigningGroupOperators = append(
-			readySigningGroupOperators,
-			srl.signingGroupOperators[memberIndex-1],
-		)
-	}
-
-	qualifiedOperators, err := retry.EvaluateRetryParticipantsForSigning(
-		readySigningGroupOperators,
-		srl.attemptSeed,
-		retryCount,
-		uint(srl.chainConfig.HonestThreshold),
+	signingLogger := logger.With(
+		zap.String("wallet", fmt.Sprintf("0x%x", walletPublicKeyBytes)),
+		zap.String("message", fmt.Sprintf("0x%x", message)),
+		zap.Uint64("signingStartBlock", startBlock),
+		zap.Uint64("signingTimeoutBlock", loopTimeoutBlock),
 	)
-	if err != nil {
-		return nil, fmt.Errorf(
-			"random operator selection failed: [%w]",
-			err,
-		)
+
+	type signerOutcome struct {
+		signature *tecdsa.Signature
+		endBlock  uint64
+		err       error
 	}
 
-	return chain.Addresses(qualifiedOperators).Set(), nil
-}
+	signerOutcomesChan := make(chan *signerOutcome, len(se.signers))
 
-// excludedMembersIndexes returns a list of excluded members' indexes for
-// the given qualified operators set.
-func (srl *signingRetryLoop) excludedMembersIndexes(
-	qualifiedOperatorsSet map[chain.Address]bool,
-	readyMembersIndexes []group.MemberIndex,
-) []group.MemberIndex {
-	includedMembersIndexes := make([]group.MemberIndex, 0)
-	excludedMembersIndexes := make([]group.MemberIndex, 0)
-	for i, operator := range srl.signingGroupOperators {
-		memberIndex := group.MemberIndex(i + 1)
-
-		if qualifiedOperatorsSet[operator] &&
-			slices.Contains(readyMembersIndexes, memberIndex) {
-			includedMembersIndexes = append(
-				includedMembersIndexes,
-				memberIndex,
+	for _, currentSigner := range se.signers {
+		go func(signer *signer) {
+			se.protocolLatch.Lock()
+			defer se.protocolLatch.Unlock()
+
```

Comment:
I am looking at `node.go` from where this code was extracted and I see this:

https://github.com/keep-network/keep-core/blob/bcc5684094a39682ab5ed52ffebbd2db24b8e80c/pkg/tbtc/node.go#L155-L168

This code was not ported to `signing.go` but I think this is fine: each sub-protocol (announce, signing, done check) performs this validation on its own.

Is this correct?

Reply:
The filter set here just says: "accept the message if it was sent by any member of the given signing group". The validation done at lower levels is more strict and says "accept the message if it was actually sent by the given member of the signing group (declared sender ID matches the signature of the operator that controls it)"